Add haproxy::instance for the creation of multiple instances of haproxy

tlimoncelli opened the pull request:
Add haproxy::instance for the creation of multiple instances of haproxy on a host.  See [MODULES-1783]

   * New Define Resource Type: haproxy::instance
   * New Define Resource Type: haproxy::instance_service
   * Class[haproxy] is now a wrapper that calls haproxy::instance
   * Add "instance" parameter to all public defines.
   * All private classes become private defined resource types.
   * Pass parameters to private defined resource types rather than
   * "inherts params" no longer used for classes that are now defines
   * Documentation updated, examples added
   * Spec tests updated and new tests added
   * Fixed: mode and options lines include an extra space

igalic commented:
oy, why does it already contain merge-conflicts :|

tlimoncelli commented:
This should merge easier if  https://github.com/puppetlabs/puppetlabs-haproxy/pull/167 is merged first.

igalic commented:
ah, yeah, i saw #167 after this one

tlimoncelli commented:
Merged.  PTAL

tlimoncelli commented:
Tests passed!  Ready for review and merge!

igalic commented on .fixtures.yml:
```diff
@@ -5,6 +5,6 @@ fixtures:
     stdlib:
       repo: "git://github.com/puppetlabs/puppetlabs-stdlib.git"
-      ref: '2.4.0'
+      ref: '2.5.1'
```
you're only bumping this here, not in manifest.json

also, with such a major rewrite, we could be more daring ;)
(but that's just my *opinion*)

tlimoncelli replied:
I'm fine with a very small increase.

igalic replied:
yes, but you're not bumping it in [metadata.json](https://github.com/StackExchange/puppetlabs-haproxy/blob/master/metadata.json#L70)

which i called manifest.json for some reason… sorry about that ~_~

tlimoncelli replied:
Right on.  I've updated metadata.json

igalic commented on files/haproxy.init:
```diff
@@ -0,0 +1,142 @@
+# pidfile: /var/run/haproxy.pid
+
+# Script Author: Simon Matter <simon.matter@invoca.ch>
+# Version: 2004060600
```
i'm surprised this passes acceptance tests on debian/ubuntu
i would think that sysv on those would complain that this is not an lsb compatible header. example from apache2 on ubuntu 14.04:

```sh
### BEGIN INIT INFO
# Provides:          apache2
# Required-Start:    $local_fs $remote_fs $network $syslog $named
# Required-Stop:     $local_fs $remote_fs $network $syslog $named
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# X-Interactive:     true
# Short-Description: Apache2 web server
# Description:       Start the web server and associated helpers
#  This script will start apache2, and possibly all associated instances.
#  Moreover, it will set-up temporary directories and helper tools such as
#  htcacheclean when required by the configuration.
### END INIT INFO
```

tlimoncelli replied:
I've decided to remove the haproxy.init file completely and instead add a parameter that let people specify an init file to be installed.  If someone is using haproxy::instance{}, they are going to have a custom init.d/haproxy-$GROUP that they want installed and if I include a sample it will just confuse them.  I could include an example in the future and make it the default to this define.

igalic replied:
another possibility is to use something like pleaserun to generate the file, but that would be a hard dependency on an external module that is still very much in flux.

igalic commented on files/haproxy.init:
```diff
@@ -0,0 +1,142 @@
+RETVAL=0
+
+start() {
+  /opt/haproxy/$BASENAME -c -q -f /etc/$BASENAME/$BASENAME.cfg
```
opt?!

igalic replied:
i'm starting to think this `file` should be a `template()`

igalic commented on manifests/config.pp:
```diff
@@ -1,34 +1,51 @@
   }
 
-  concat { $haproxy::config_file:
+  if $config_dir != undef {
```
that seems… a weird semantic.
you'd have to set `config_dir` explicitly as `undef`.
but, what does that mean?

tlimoncelli replied:
This is a private define. The caller will always set config_dir if there is a parent directory that needs to be created.  Otherwise it is left undef and is not created.

igalic commented on manifests/instance.pp:
```diff
@@ -0,0 +1,232 @@
+    $_global_options = $haproxy::params::global_options
+  } else {
+    $_global_options = $global_options
+  }
```
this whole bit reads very confusing.
what's the reason we can't just `include haproxy::params` regardless before we dive down into the `if`s?

tlimoncelli replied:
If all of $global_options, $default_options and $config_file are defined, doing the "include haproxy::params" is a waste of time.  That said, I may be guilty of premature optimization.  Would you recommend I simplify the code and do one "include haproxy::params" early in the class?

igalic commented on manifests/instance_service.pp:
```diff
@@ -0,0 +1,89 @@
+#
+define haproxy::instance_service (
+  $haproxy_package = 'haproxy',
+  $bindir = '/opt/haproxy',
```
this seems like a bad default for most installations ;)

igalic commented on manifests/instance_service.pp:
```diff
@@ -0,0 +1,89 @@
+#   If you have an instance called "foo", you know "haproxy-foo"
+#   is a link to the binary it should be using.
+# * Create an init.d file named after the instance. This way
+#   Service[$instance] can start/restart the service.
```
why?

tlimoncelli replied:
Short answer: That's how we do it at Stack.

Longer answer:   As the next paragraph says, this module is just one example of how to do this. Other sites can choose to do it other ways.  All you have to do is set up Service[$instance_name], and there are many other ways is can be done.

igalic commented on manifests/peers.pp:
```diff
@@ -1,11 +1,25 @@
+    $instance_name = "haproxy-${instance}"
+    #$config_dir = inline_template($haproxy::params::config_dir_tmpl)
+    $config_file = inline_template($haproxy::params::config_file_tmpl)
+  }
```
i'm starting to think that this whole if $instance == haproxy { } else {} pattern needs to be collapsed / contained better.

tlimoncelli replied:
I don't like repeating the same 10 lines in multiple files.  I couldn't come up with any other way to do this.

It can't be part of haproxy::params.pp because the value of $instance changes. 

If someone suggests a better way to do it, I'm willing to try anything!

igalic replied:
i was thinking a function, but i'm not sure… maybe @hunner or @mhaskel or someone else from puppetlabs have an idea?

tlimoncelli commented:
PTAL

igalic commented:
@tlimoncelli i've done my best to review your code, and the *only* thing left is for someone from puppetlabs to chime in. Well, that and a [rebase](http://git-scm.com/book/en/Git-Branching-Rebasing)

cyberious commented on manifests/backend.pp:
```diff
@@ -57,17 +57,30 @@
+    $instance_name = "haproxy-${instance}"
+    #$config_dir = inline_template($haproxy::params::config_dir_tmpl)
+    $config_file = inline_template($haproxy::params::config_file_tmpl)
+  }
```
Please remove commented lines and also prefix private variables to have $_

cyberious commented on manifests/balancermember.pp:
```diff
@@ -93,14 +93,27 @@
+    $instance_name = "haproxy-${instance}"
+    #$config_dir = inline_template($haproxy::params::config_dir_tmpl)
+    $config_file = inline_template($haproxy::params::config_file_tmpl)
+  }
```
Again commented lines should be removed.

hunner commented:
Hello. The nature of puppetlabs modules is that they are patterned after industry standards and conventions and are supported as such by the module team. The module team believes that multiple haproxy instances on a single node does not follow industry convention. Merging this PR would mean that we must maintain the functionality for questionable benefit to the majority of users. I hope you understand our concerns.

tlimoncelli commented:
As you can guess, I'm disappointed by this response and would  like to discuss it further before the matter is considered closed.

This is following industry convention.  This is a configuration that is recommended by Willy, the author of haproxy, for anyone that needs to isolate frontends from each other or create separate domains of reliability.  The conventions used are similar to other puppet modules that permit multiple instances. For example the puppetlabs/tomcat module. Frankly, it is a tautology to say that Puppet users don't do this: they can't because the module doesn't support it.

The changes are minor and fully backwards compatible.  The classes are carefully constructed to be fully backwards compatible, as demonstrated by the spec tests as well as personal experience. If this PR is merged, I volunteer to work with anyone that finds incompatibilities until the next major release, and will create spec tests to demonstrate that they are fixed.

The one file that may be slightly difficult to support is instance_service.pp because it is a demonstration of how a site might use these new features but sites may choose to use it differently. If you prefer, I can move the contents of this file to the documentation as an example of how one might create the Service[] instance so-as to not to imply that this is the only way to use the module. This would be sad, considering that it is a fully working example with spec tests. I would prefer to alter the documentation to be more clear that this is an example, possibly renaming it to service_example.pp.

Lastly, I'm particularly surprised because I checked with the community before moving forward with this project.  I created (MODULES-1783) and Morgan Haskel wrote "that sounds like a reasonable feature to add".

I'd be glad to discuss this further.

Sincerely,
Tom Limoncelli

hunner commented:
Could you rebase this an add some acceptance tests?

tlimoncelli commented:
I've rebased and added some tests.  Sadly TravisCI fails the tests.  The tests all work on my machine.  It seems to have something to do with the "resource" function.  Any suggestions?

TomOnTime commented:
All problems found and fixed.

igalic commented:
@TomOnTime could you please do a proper [rebase](http://git-scm.com/book/en/Git-Branching-Rebasing)? and [squash down](http://www.git-scm.com/book/en/Git-Tools-Rewriting-History#Squashing-Commits) the commits that are failing tests!

tlimoncelli commented:
Done.  PTAL

igalic commented:
![image of commits](http://i.imgur.com/gjJ46yj.png) still has the merge commit…

when you…

```
git remote add upstream git@github.com:puppetlabs/puppetlabs-haproxy
git remote update
git rebase upstream/master
```

it *should* vanish… ***should***…


git :(

TomOnTime commented:
When github says the code needs to be rebased, it should include a link to those 3 commands.  I've been doing it a much longer, and more painful, way.

igalic commented:
\o/

you're perfectly welcome!
***
why does this contain merge conflicts already again :(

TomOnTime commented:
I think I squashed things properly now. (I have a love-hate relationship with Git).  PTAL

tphoney commented:
you need to push this change, then it will be reflected in github. It is a pain, i agree.

tlimoncelli commented:
@tphoney Please pretend I'm new to Git and tell me what you want me to do.

igalic commented:
<blockquote class="twitter-tweet" lang="en"><p>git people trying to use hg: &quot;WHY WON&#39;T IT LET ME DO THIS&quot;<br/>hg people trying to use git: &quot;WHY DID IT LET ME DO THAT&quot;</p>&mdash; ❀❁ eevee ❁❀ (@eevee) <a href="https://twitter.com/eevee/status/587789870103298048">April 14, 2015</a></blockquote> <script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

tphoney commented:
THE PAIN, IT HURTS. I used SVN for many years, although not great, i could do things. Even Perforce was ok. With Git, i know i can do it, it feels like magic if it does work. Most of the time i trawl the internet for solutions, and weep.

The following command will print out your commits. 
git log --pretty=oneline
To squash (the last 2 commits)
git rebase --interactive HEAD~2
Then read the text squash the commits you want to, and save. 
(You can view the squashed commit  with the log command above "git log --pretty=oneline")
then push it to github (you will push to the branch you were working on)
git push -f origin destination_branch

Or read this
http://stackoverflow.com/questions/2563632/how-can-i-merge-two-commits-into-one

mhaskel commented:
@tlimoncelli this will also need updates in the README for new parameters and new classes/defines. Thanks!

bmjen commented:
Also, acceptance tests are failing for this branch.

TomOnTime commented:
@mhaskel  Docs updated.  Thanks for the reminder!
@bmjen Would you please link to the failing test?  I see all green here: https://travis-ci.org/puppetlabs/puppetlabs-haproxy/builds/59799099

bmjen commented:
@tlimoncelli Here is a snippet of the failures from my acceptance test run on centos-6.
https://gist.github.com/bmjen/9ddcd4cd54093cf70f71

Also failed similarly in Ubuntu-1404.

bmjen commented:
@tlimoncelli Also, the travis-ci you linked is for the PR passing unit tests.

tlimoncelli commented:
@bmjen Ah, I missed that you said acceptance test.  Thanks, now I get it.  Well, my intention is that this big change should be fully backwards compatible, so I'd like to help get those tests passing.  Are they tests I can run some how so I can reproduce the issue locally?  Failing that, could we do a hangout/skype/whatever?

bmjen commented:
@tlimoncelli if you have a vagrant/vbox setup locally, you can run the acceptance tests via `bundle exec rspec spec/acceptance`

tlimoncelli commented:
@bmjen I can't get the acceptance tests working on the master branch, let alone the multiple_instances branch.  I doubt I have the right hardware to run them.

bmjen commented:
@tlimoncelli I can help look at the failures if time permits. What errors are you getting when trying to run `bundle exec rspec spec/acceptance`?

tlimoncelli commented:
Here's what I've tried so far:
https://gist.github.com/tlimoncelli/0dc02b5c5dd614d0e7f5

juniorsysadmin commented:
Perhaps the commands that you were after were `bundle install` then `bundle exec rake beaker`. You can also choose which vagrant image from spec/acceptance/nodesets to run the acceptance tests on with eg. `BEAKER_set=ubuntu-server-1404-x64 bundle exec rake beaker`

I would recommend jumping onto the #puppet IRC channel on Freenode, and even if you don't get a response immediately you'll get pointed in the right direction a bit faster.

tlimoncelli commented:
I finally built an environment that could run the acceptance tests.  They all pass now.
I've updated the docs, rebased, etc.  Please take a look and consider merge if all is good.

TomOnTime commented:
@bmjen Once I finally got an environment that could run vbox and the other things that the acceptance tests require, it all worked. Can you try again and let me know if you see any issues still?

tlimoncelli commented:
@hunner @igalic @mhaskel PTAL.

tphoney commented:
Thanks @tlimoncelli for your work, but can't be merged as it has conflicts. Please rebase them on the current master, fix the conflicts and repush here. https://git-scm.com/book/en/v2/Git-Branching-Rebasing

tlimoncelli commented:
Will do.

mentat commented:
Hi @tlimoncelli I'd love to see this PR get merged, can you resolve the merge conflicts so I can test?

TomOnTime commented:
@mentat I'm struggling with git, but I will try to get this done soon.

igalic commented:
@TomOnTime we offer git related hand-holding in #puppet-community on freenode
(among other things :)

tlimoncelli commented:
rebased. PTAL

igalic commented:
@tlimoncelli almost! you still got a remote-tracking branch commit in there.
Please refer to my https://github.com/puppet-community/puppet-jira/pull/99#issuecomment-152528108 for how to do this cleanly.

TomOnTime commented:
Thanks for the link.  I got it this time!

igalic commented:
@TomOnTime \o/

tlimoncelli commented:
What do I do now?

igalic commented:
now we wait…
![wait](https://media.giphy.com/media/u5eXlkXWkrITm/giphy.gif)
for someone from @puppetlabs to re-review this pr, or just press that sweet, sweet green merge button.

mentat commented:
I'll run acceptance on this today.

tphoney commented:
cool !!

tlimoncelli commented:
FYI:
This includes https://github.com/puppetlabs/puppetlabs-haproxy/pull/175 merged in already. Hopefully they'll both be merged.

hunner commented:
:+1:

igalic commented:
\o/

TomOnTime commented:
\o/    Happy Dance

Thanks, y'all!
